File: app/src/main/java/ua/hospes/nfs/marathon/data/drivers/DriversRepositoryImpl.java
package ua.hospes.nfs.marathon.data.drivers;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import ua.hospes.nfs.marathon.data.drivers.mapper.DriversMapper;
import ua.hospes.nfs.marathon.data.drivers.storage.DriversDbStorage;
import ua.hospes.nfs.marathon.data.team.models.TeamDb;
import ua.hospes.nfs.marathon.data.team.storage.TeamsDbStorage;
import ua.hospes.nfs.marathon.domain.drivers.DriversRepository;
import ua.hospes.nfs.marathon.domain.drivers.models.Driver;

/**
 * @author Andrew Khloponin
 */
@Singleton
public class DriversRepositoryImpl implements DriversRepository {
    private final DriversDbStorage dbStorage;
    private final TeamsDbStorage teamsDbStorage;


    @Inject
    public DriversRepositoryImpl(DriversDbStorage dbStorage, TeamsDbStorage teamsDbStorage) {
        this.dbStorage = dbStorage;
        this.teamsDbStorage = teamsDbStorage;
    }


    @Override
    public Observable<Driver> get() {
        return dbStorage.get()
                .flatMap(driverDb -> Observable.zip(Observable.just(driverDb), getTeamById(driverDb.getTeamId()), DriversMapper::map));
    }

    @Override
    public Observable<Driver> get(int... ids) {
        return dbStorage.get(ids)
                .flatMap(driverDb -> Observable.zip(Observable.just(driverDb), getTeamById(driverDb.getTeamId()), DriversMapper::map));
    }

    @Override
    public Observable<Driver> getByTeamId(int teamId) {
        return dbStorage.getTeamById(teamId)
                .flatMap(driverDb -> Observable.zip(Observable.just(driverDb), getTeamById(driverDb.getTeamId()), DriversMapper::map));
    }

    @Override
    public Observable<List<Driver>> listen() {
        return dbStorage.listen()
                .flatMap(driverDbs -> Observable.from(driverDbs)
                        .flatMap(driverDb -> Observable.zip(Observable.just(driverDb), getTeamById(driverDb.getTeamId()), DriversMapper::map))
                        .toList());
    }


    @Override
    public Observable<Boolean> save(Driver driver) {
        if (driver.getId() == -1) {
            return dbStorage.add(DriversMapper.map(driver)).map(result -> result.getResult() != 0);
        } else {
            return dbStorage.update(DriversMapper.map(driver)).map(result -> result.getResult() != 0);
        }
    }

    @Override
    public Observable<Boolean> delete(Driver driver) {
        return dbStorage.remove(DriversMapper.map(driver)).map(count -> count != 0);
    }

    @Override
    public Observable<Void> clear() {
        return dbStorage.clear();
    }


    private Observable<TeamDb> getTeamById(int id) {
        return teamsDbStorage.get().singleOrDefault(null, team -> id == team.getId());
    }
}

File: app/src/main/java/ua/hospes/rtm/data/DriversRepositoryImpl.kt
package ua.hospes.rtm.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import ua.hospes.rtm.db.drivers.DriverDAO
import ua.hospes.rtm.db.drivers.toDomain
import ua.hospes.rtm.db.team.TeamDAO
import ua.hospes.rtm.domain.drivers.Driver
import ua.hospes.rtm.domain.drivers.DriversRepository
import ua.hospes.rtm.domain.drivers.toDbEntity

internal class DriversRepositoryImpl(private val dao: DriverDAO, private val teamDAO: TeamDAO) : DriversRepository {
    override suspend fun get(): List<Driver> =
            withContext(Dispatchers.IO) { dao.get().map { it.toDomain(teamDAO) } }

    override suspend fun get(vararg ids: Long): List<Driver> =
            withContext(Dispatchers.IO) { dao.get(*ids).map { it.toDomain(teamDAO) } }

    override suspend fun getByTeamId(teamId: Long): List<Driver> =
            withContext(Dispatchers.IO) { dao.getByTeamId(teamId).map { it.toDomain(teamDAO) } }

    override suspend fun getNotInRace(teamId: Long): List<Driver> =
            withContext(Dispatchers.IO) { dao.getNotSelected(teamId).map { it.toDomain(teamDAO) } }

    override fun listen(): Flow<List<Driver>> = dao.observe().map { list -> list.map { it.toDomain(teamDAO) } }


    override suspend fun addDriversToTeam(teamId: Long, vararg driverIds: Long) =
            withContext(Dispatchers.IO) { dao.addDriversToTeam(teamId, driverIds) }

    override suspend fun removeDriversFromTeam(teamId: Long) =
            withContext(Dispatchers.IO) { dao.removeDriversFromTeam(teamId) }


    override suspend fun save(driver: Driver): Driver =
            withContext(Dispatchers.IO) { dao.save(driver.toDbEntity()).let { driver.copy(id = it) } }

    override suspend fun delete(id: Long) =
            withContext(Dispatchers.IO) { dao.delete(id) }

    override suspend fun clear() =
            withContext(Dispatchers.IO) { dao.clear() }
}